open-apparel-registry/open-apparel-registry
# Create facility list items table interactions

## Overview

This PR accomplishes a few things:

- updates the API to handle confirming or rejecting potential facility matches
- adds tests for API confirm/reject logic
- updates the UI to add several interactions to the facility list items list table, including:
   - showing toggleable detail rows for list items in the matched or error states, along with additional details
    - showing a confirm/reject match UI along with a confirmation dialog to use for confirming or rejecting matches
- adds links to a user's facilities on the main list page and also in the navigation bar dropdown
- update the match functionality to create some random matches if the list item id % 4 === 0. This is done because the previous naive matching function was setting everything with 100% confidence

Connects #192

## Demo

![confirm-match](https://user-images.githubusercontent.com/4165523/53187012-2adb5e00-35d0-11e9-9171-ba62f2d0028a.gif)

![reject-matches](https://user-images.githubusercontent.com/4165523/53187023-2f077b80-35d0-11e9-888a-a96e06352e15.gif)

## Testing

- get this branch
- resetdb
- serve the app, then register for and log in with a new account
- upload a list from the set of lists committed to the repo (pick one of the larger ones to ensure lots of matches)
- in the VM, run the following commands to process the list:

```
./scripts/manage batch_process --action parse --list-id 15
./scripts/manage batch_process --action geocode --list-id 15
./scripts/manage batch_process --action match --list-id 15
```

- visit your lists page in the UI, then click on the list to go to the details page
- try the following interactions:
   - clicking on an ERROR or MATCH or CONFIRMED_MATCH row should expand that row to display details, and close all other non-CONFIRM details rows
   - clicking on an expanded non-CONFIRM detail row should cause it to collapse back to a normal table row
    - confirming a match through the UI should update that table row to display it as a CONFIRMED_MATCH with a link to the facility that has matched
    - rejecting one of several matches should keep the confirm row open, but show the rejected match as rejected
    - rejecting the last potential match for a list item should create a new facility then update the details row to show it in CONFIRMED_MATCH state

- run `./scripts/test` to verify that they pass
